File: medium/linkedlist/lru.cpp
#include <unordered_map>
#include <iostream>
using namespace std;
class LRUCache {

    struct Node {
        Node* next;
        Node* prev;
        int val;
        int key;
        Node(int key, int val) {
            this->val = val;
            this->prev = nullptr;
            this->next = nullptr;
            this->key = key;
        }

    };


    public:
        int capacity;
        int size;
        Node* head; // least recently used
        Node* tail; // most recently used
        unordered_map<int, Node*> mp;

        LRUCache(int capacity) {   
            this->size = 0;
            this->capacity = capacity;
            head = nullptr;
            tail = head;
        }

        void bringToFront(Node* cur, int key) {
            if (tail == cur) return; // already at front
            Node* prev = cur->prev;
            Node* next = cur->next;
            // case: prev exists
            if (prev) prev->next = next;
            else head = next;

            // case: next exists
            if (next) next->prev = prev;

            // update head
            if (head) head->prev = nullptr;

            tail->next = cur;
            cur->prev = tail;
            cur->next = nullptr;
            tail = cur;
        }
        
        int get(int key) {
            if (mp.find(key) == mp.end()) {
                cout << "get(" << key << ") is nullptr" << endl;
                return -1;
            } else {
                // bring to front; Most recently used
                Node* cur = mp[key];
                bringToFront(cur, key);
                return cur->val;
                
            }
            
        }
        
        void put(int key, int value) {
            if (size == 0) {
                Node* cur = new Node(key, value);
                mp[key] = cur;
                size++;
                head = cur;
                tail = cur;

            } 
            else if (size < capacity) {
                if (mp.find(key) == mp.end()) {
                    Node* cur = new Node(key, value);
                    mp[key] = cur;
                    size++;
                    // add to end of LL
                    tail->next = cur;
                    cur->prev = tail;
                    tail = cur;
                } else if (mp.find(key) != mp.end()) {
                    // if key exists
                    Node* cur = mp[key];
                    cur->val = value;
                    // bring to front; Most recently used
                    bringToFront(cur, key);
                    return;
                    
                }
            } else if (size == capacity) {
                if (head == tail) {
                    // if capacity is 1, delete entries
                    mp.erase(head->key);
                    delete head;

                    // create entry
                    head = new Node(key, value);
                    mp[key] = head;
                    tail = head;
                    return;
                }
                // logic to replace Least recently used
                
                // first check if key exists
                if (mp.find(key) != mp.end()) {
                    // if key exists
                    Node* cur = mp[key];
                    cur->val = value;
                    // bring to front; Most recently used
                    bringToFront(cur, key);
                    return;
                    
                }


                cout << "MAX capacity" << endl;
                // remove head (LRU) & update map
                Node* temp = head;
                head = head->next;
                if (head) head->prev = nullptr;
                mp.erase(temp->key); // IMPORTANT
                delete temp;
                size--;
                
                // push new Node to end (MRU)
                Node* cur = new Node(key, value);
                mp[key] = cur;
                tail->next = cur;
                cur->prev = tail;
                tail = cur;
                size++;

            }

        }
    };
    

File: medium/linkedlist/lru.py
"""
linkedlist to keep track of most recently/least recently used with head and tail
map to get in O(1) where our nodes are for a key (key -> node)

"""
class Node:
    def __init__(self, key, val,prev=None, next=None):
        self.key = key
        self.val = val
        self.prev = prev
        self.next = next

class LRUCache:

    def __init__(self, capacity: int):
        self.m = {} # map key -> node
        self.capacity = capacity

        self.head = None
        self.tail = None
        

    def get(self, key: int) -> int:
        if key in self.m:
            node = self.m[key]

            # move it up to head (most recently used)
            self.remove(node)
            self.add(node)
            return node.val
        else:
            return -1
        

    def put(self, key: int, value: int) -> None:
        if key in self.m:
            # upd node
            node = self.m[key]
            node.val = value
            
            # bring it to head O(1)
            self.remove(node)
            self.add(node)
        else:
            node = Node(key, value)
            self.m[key] = node
            self.add(node)

            #evict if capacity
            if len(self.m) > self.capacity:
                del self.m[self.tail.key]
                self.remove(self.tail)
    
    # new funcs
    def add(self, node):
        # add node to head of ll
        node.prev = None
        node.next = self.head
        if self.head:
            self.head.prev = node
        self.head = node
        if not self.tail:
            self.tail = node
    def remove(self, node):
        if node.prev:
            node.prev.next = node.next
        else:
            # removing head
            self.head = node.next
        if node.next:
            node.next.prev = node.prev
        else:
            # node is tail
            self.tail = node.prev # removed
    

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)